### ISSUE TITLE
RFC: `approx_count_distinct` state table encoding

### ISSUE THREAD

**Issue by jon-chuang** [singularity-data/risingwave#3414]
# RFC Template

## Background

We need to consider `approx_count_distinct` state table encoding to facilitate:
1. Recovery from checkpoint upon failure
2. Recovery from checkpoint upon config change actor placement

### Encoding of `approx_count_distinct`

key: `table_id | bucket_id (16 bits) | count_index (8 bits)` (count_index ranges from 0-48 and represents trailing zeroes)
value: `count (u64)`

Seems like vnode does not factor into the encoding since we are assuming that `approx_count_distinct` is a simple agg singleton operator.

## Future Optimizations

We could consider `approx_count_distinct` with `group_by`. Because of the large default in-memory state size of `approx_count_distinct`, we may only be able to support a small number of `group_by` keys.

**Comment by fuyufjh** [comment]
An alternative design is to store all counts of a bucket in one KV-entry, the size would be 92 bytes and seems okay to me.

**Comment by fuyufjh** [comment]
I am also thinking about whether [my design](https://github.com/singularity-data/risingwave/issues/2727#issuecomment-1146557868) is good, because the memory usage it quite high (~1.5MB). This would be a problem for 2-phase agg or HashAgg.

**Comment by jon-chuang** [comment]
Hmm, indeed. 

One way is to use the `bucket_index` to calculate vnode for 2-phase agg. So the state size is split n-way amongst n actors.

The partial agg can compute partial results of the harmonic mean.

**Comment by fuyufjh** [comment]
> One way is to use the bucket_id as distribution key to calculate vnode for 2-phase agg. So the state size is split n-way amongst n actors.

That's a very interesting insight! But it seems hard to be used along with other agg functions.

**Comment by fuyufjh** [comment]
> As for hash agg, we may not need to support it. Or, we could throw a runtime error if the number of group by keys grows too large?

Yes, this is a feasible solution. I am just not sure whether this is what our users want. Anyway, the `approx_count_distinct` is expected to be a lightweight (but approximate) alternative for `count(distinct)`, but it seems not lightweight and even cause more limitations.

**Comment by fuyufjh** [comment]
We may hold this RFC for a while, I think

**Comment by jon-chuang** [comment]
> but it seems not lightweight and even cause more limitations.

Yes, indeed. It is not lightweight when the distinct elements are small in number and group keys are large in number. I think we may need the sparse-dense transition to handle this.

**Comment by fuyufjh** [comment]
I have added back this issue to the project board. We may need to make a decision on whether we decide to support or disable `approx_count_distinct`, especially the non-append-only one.

**Comment by fuyufjh** [comment]
The difficulty is: the state, as described in the RFC, is very special compared with these aggregate functions that we have supported (e.g. min/max or sum/count)

**Comment by jon-chuang** [comment]
Note that one has to persist a different state per vnode.

**Comment by xxchan** [comment]
What's the status of this RFC?

**Comment by stdrc** [comment]
> What's the status of this RFC?

Current status of `approx_count_distinct` is:

* It can work on append-only stream, for both hash agg and simple agg. The state is encoded into a condensed int array as one datum in the "intermediate state table". The encoding/decoding process is not that efficient. I haven't compare the performance with normal `count(distinct x)` though.
* It has an implementation for non-append-only stream, can count correctly if no recovery is involved. By recovery, I mean, actor rebuilding for simple agg, and, group cache eviction or actor rebuilding for hash agg. It encode its state by simply store the result value in the "intermediate state table", and when recovering, it simply use that result as "initial count" which is summed with the HLL result for new inputs.

As said in previous discussion, we may consider not to allow this function in hash agg, because the state can explode ridiculously. When there's no many rows within each group, with a correctly implemented non-append-only state, there're fixed 65536 rows for each group, and the value column is of type `int64[]`. At the same time, a normal `count(distinct x)` requires only M rows in the state where M is the number of distinct values of `x`, with value column to be a simple `int64` as the duplicate count of each distinct value. It seems not reasonable to use `approx_count_distinct` anyway when doing a grouped aggregation.

If we do decide to only support this agg for simple agg, we can fearlessly have a state table storing each bucket as one row, then we can have a correct non-append-only implementation.
